### PR TITLE
(concurrentbatchprocessor): add in-flight byte limit to producer queue

### LINKED ISSUE #11: Support for blocking OTel-Arrow pipeline
The `otelarrow` exporter uses the `exporterhelper`, a common component from OpenTelemetry Collector-core. The `exporterhelper` does not support back-pressure, instead of drops data or returns immediately to the user.

We are tracking these issues:
https://github.com/open-telemetry/opentelemetry-collector/issues/8245
https://github.com/open-telemetry/opentelemetry-collector/issues/8122

### LINKED ISSUE #80: Develop a concurrent memory-limited batch processor
Briefly: we have had trouble tuning the configuration of an OTel-Arrow pipeline so that data does not drop. The root cause of this is the `exporterhelper` configuration and concurrency behavior combined with the core `batchprocessor` configuration and concurrency behavior.

We will develop an alternative based on the the core `batchprocessor` by forking the code, and then:

1. Modify the batchprocessor to block requests until the batch response is received
2. Modify the batchprocessor to return error or no-errors to the producer
3. Modify the batchprocessor to send >1 request in parallel, whenever data is available, using background goroutines for each batch
4. Establish an in-flight-bytes limit that prevents more than N bytes from entering its internal queue, to limit the amount of concurrent work and overall resource usage by the rest of the pipeline.

### PR DESCRIPTION
**Pull request by moh-osman3**
This PR adds a new config option in the `concurrentbatchprocessor` named `max_in_flight_bytes`. This config option will be used by processor to control admission (using a semaphore) into the producer queue based on the size of requests.

Fixes #80.
Fixes #11.

### CODE REVIEW

File: collector/processor/concurrentbatchprocessor/batch_processor.go [hunk 1]
```diff
@@ -565,7 +565,12 @@ func (bt *batchTraces) sizeBytes(data any) int {
 	return bt.sizer.TracesSize(data.(ptrace.Traces))
 }
 
-func (bt *batchTraces) export(ctx context.Context, req any) error {
+func (bt *batchTraces) export(ctx context.Context, req any) (retErr error) {
+	defer func() {
+		if r := recover(); r != nil {
+			retErr = errors.New(r.(string))
+		}
+	}()
```
Comment (jmacd) [inline collector/processor/concurrentbatchprocessor/batch_processor.go]: I'd like to see a new top-level function--
```
func recoverError(retErr *error) {
		if r := recover(); r != nil {
			*retErr = errors.New(fmt.Sprint(r))
		}
	}
```

Also note, the `r` value will be whatever is passed to `panic`, so not always a string. Suggest letting `fmt.Sprint` figure out the best to-string method.

Comment (jmacd) [reply collector/processor/concurrentbatchprocessor/batch_processor.go]: (So, `fmt.Errorf("%v", r)` also works, no need for `errors.New()` then.)

Comment (moh-osman3) [reply collector/processor/concurrentbatchprocessor/batch_processor.go]: Thanks for the review! Fixed

**Review by jmacd — APPROVED** [review]
Thanks for iterating on the tests with me! 🚀

**Pull request merged**